components/availability/src/Availability.svelte

```diff
@@ -646,6 +723,189 @@
+  // Note some fun behaviour: If you drag a long event overtop a busy spot, it will automatically split it into multiple events.
+  // You can set partial_bookable_ratio to 0 in order to prevent this / let you book atop busy spots.
+  let mouseIsDown = false;
+  let dragStartSlot: SelectableSlot | null = null;
+  let dragStartDay: Day | null = null;
+  let draggingExistingBlock: boolean = false;
+
+  function resetDragState() {
+    days.forEach((day) =>
+      day.slots
+        .filter((x) => x.selectionPending)
+        .forEach((x) => {
+          x.selectionPending = false;
+        }),
+    );
+
+    mouseIsDown = false;
+    dragStartDay = null;
+    dragStartSlot = null;
+    draggingExistingBlock = false;
+    draggedBlock = null;
+  }
+
+  // draggedBlock is a member of the SortedSlots array; it has an arbitrary number of timeSlots and is represented by a single start_time and end_time
+  // draggedBlockSlots is an array of timeSlots that fall within the draggedBlock's start_time and end_time
+  let draggedBlock: SelectableSlot | null = null;
+  $: draggedBlockSlots = draggedBlock
+    ? dragStartDay?.slots.filter(
+        (slot) =>
+          slot.start_time >= draggedBlock!.start_time &&
+          slot.end_time <= draggedBlock!.end_time,
+      )
+    : [];
+
+  function startDrag(slot: SelectableSlot, day: Day) {
+    if (allow_booking) {
+      // Retain the initially-clicked slot and day, so we can adjust if you've moved across dates, etc.
+      dragStartSlot = slot;
+      dragStartDay = day;
+
+      if (slot.selectionStatus === SelectionStatus.SELECTED) {
+        draggingExistingBlock = true;
+        draggedBlock =
+          sortedSlots.find(
+            (block) =>
+              slot.start_time >= block.start_time &&
+              slot.end_time <= block.end_time,
+          ) || null;
+      } else if (slotSelection.length < max_bookable_slots) {
+        slot.selectionPending = true;
+      }
+    }
+  }
+
+  function addToDrag(slot: SelectableSlot, day: Day) {
+    if (mouseIsDown) {
+      if (draggingExistingBlock && dragStartSlot && dragStartDay) {
+        // dragStartSlot && dragStartDay are superfluous here, but type errors are thrown if we don't explicitly check for them
+
+        // Measure the distance between where you started and where you've dragged to, vertically.
+        let delta =
+          day.slots.indexOf(slot) - dragStartDay.slots.indexOf(dragStartSlot);
+
+        // Remove "pending" status from your previous location
+        days.forEach((day) =>
+          day.slots
+            .filter((slot) => slot.selectionPending)
+            .forEach((slot) => (slot.selectionPending = false)),
+        );
+
+        draggedBlockSlots?.forEach((slot) => {
+          let slotPlusDelta =
+            day.slots[dragStartDay!.slots.indexOf(slot) + delta];
+          if (slotPlusDelta) {
+            slotPlusDelta.selectionPending = true;
+          }
+        });
+
+        days = [...days]; // re-render
```

Comment:
is this necessary since `days` is a reactive variable?

Reply:
It weirdly is! https://svelte.dev/tutorial/updating-arrays-and-objects